src/jit/compiler.h

```diff
@@ -911,6 +911,26 @@ class LclVarDsc
 
     var_types lvaArgType();
 
+    // Returns true if this variable contains GC pointers (including being a GC pointer itself).
+    bool HasGCPtr()
+    {
+        return varTypeIsGC(lvType) || ((lvType == TYP_STRUCT) && m_layout->HasGCPtr());
+    }
+
+    // Returns the layout of a struct variable.
+    ClassLayout* GetLayout()
+    {
+        assert(varTypeIsStruct(lvType));
+        return m_layout;
+    }
+
+    // Sets the layout of a struct variable.
+    void SetLayout(ClassLayout* layout)
+    {
+        assert(varTypeIsStruct(lvType));
+        m_layout = layout;
```

Comment:
Heh, I was just working on finishing up this PR and this was something I was intending to put in comments. Yes, we should definitely avoid the `lvVerTypeInfo` handle as it's duplicate and duplicate stuff is always at risk of getting out of sync and cause problems.

I actually did this but decided to keep it our of this PR to avoid some weirdness that goes on in `lvaSetStruct` - it has a `setTypeInfo` parameter that controls the setting of the handle in `lvVerTypeInfo` independently of layout. It would seem that when verification is enabled the handle in `lvVerTypeInfo` can perhaps be somehow different. I need to double check what's going on, though verification is AFAIK not a thing in .NET Core.

I also run into an annoyance caused by the implicit byref transformation - I would prefer to keep the `assert(varTypeIsStruct(lvType));` asserts in `Set/GetLayout` but during that transformation the `LclVarDsc` is temporarily in a weird state where it no longer has struct type but the handle from layout is needed. For this and other reasons I suspect that I may have to clean up the implicit byref transformation first.

Reply:
Only high level - we likely need to move this transform out of global morph, in front of it. The temporary weird state of the `LclVarDsc` might be unavoidable but I'd fell more comfortable if it's dealt with before global morph.

This would also tie in with #20957. Overall I think it should work like this:
* `LocalAddressVisitor` deals with all indirect lclvar access and produces `GT_LCL_VAR`/`GT_LCL_FLD` whenever possible. When not possible the lclvar is left address exposed. It should also leave some breadcrumbs to drive promotion (e.g. collect reference counts).
* Promotion of `LclVarDsc`s
* Transform IR according to promotion decisions above. Implicit byref transform should happen at the same time (since it's related to promotion).
* Global morph - at this point most, if not all, lclvar related transforms have already been applied. Morph has less work to do and we avoid some chicken & egg situations (e.g. having to kill assertions due to promotion).
* No more demotion of implicit by ref args

This does require an extra traversal of the IR so it would probably have a small throughput hit. But I keep finding ways to improve throughput in other parts of the JIT so I'm not worried about that.